File: root/src/DataCollect.R
require(R6)
require(arrow)
require(readxl)
require(tidyverse)

DataCollect <- R6Class(
  "DataCollect",
  
  public = list(
    
    tickers       = NULL,
    root_path     = NULL,
    repo_path     = NULL,
    note_path     = NULL,
    data_path     = NULL,
    raw_path      = NULL,
    bbg_px_path   = NULL,
    bbg_bond_path = NULL,
    
    initialize = function(){
      
      if ("rstudioapi" %in% installed.packages() && rstudioapi::isAvailable()) {
        wd <- dirname(rstudioapi::getActiveDocumentContext()$path)
      } else {
        wd <- getwd()
      }
      
      setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
      
      self$tickers <- c("ANGL", "FALN", "JNK", "HYG")
      
      self$root_path     <- dirname(getwd())
      self$repo_path     <- dirname(self$root_path)
      self$note_path     <- file.path(self$root_path, "notebooks")
      self$data_path     <- file.path(self$repo_path, "data")
      self$raw_path      <- file.path(self$data_path, "RawData")
      
      
      self$bbg_px_path   <- "/Users/diegoalvarez/Desktop/BBGData/data"
      if (file.exists(self$bbg_px_path) == FALSE){
        self$bbg_px_path  <- "C:\\Users\\Diego\\Desktop\\app_prod\\BBGData\\data"
        }
      
      self$bbg_bond_path  <- "/Users/diegoalvarez/Desktop/BBGData/ETFIndices/BondPricing"
      if (file.exists(self$bbg_bond_path) == FALSE){
        self$bbg_bond_path <- "C:\\Users\\Diego\\Desktop\\app_prod\\BBGData\\ETFIndices\\BondPricing"
        }
      
      if (!dir.exists(self$note_path)){dir.create(self$note_path, recursive = TRUE)}
      if (!dir.exists(self$data_path)){dir.create(self$data_path, recursive = TRUE)}
      if (!dir.exists(self$raw_path)) {dir.create(self$raw_path, recursive = TRUE)}
      
    },
    
    get_px = function(){
      
      paths <- lapply(self$tickers, function(path) file.path(self$bbg_px_path, paste0(path, ".parquet")))
      
      df_tmp <- do.call(rbind, lapply(paths, function(p) read_parquet(p))) %>% 
        select(date, security, "PX_last" = value) %>% 
        mutate(date = as.Date(date))
      
      return(df_tmp)
      
    },
    
    get_bond_attributes = function(){
      
      paths <- lapply(self$tickers, function(path) file.path(self$bbg_bond_path, paste0(path, ".parquet")))
      keep_cols <- c("date", "security", "variable", "value")
      
      df_tmp <- do.call(rbind, lapply(paths, function(p) read_parquet(p, col_select = keep_cols))) %>% 
        pivot_wider(names_from = variable, values_from = value) %>% 
        mutate(date = as.Date(date)) %>% 
        rename(
          "yas_yld"    = "YAS_BOND_YLD",
          "mod_dur"    = "YAS_MOD_DUR",
          "yas_spread" = "YAS_YLD_SPREAD",
          "gspread"    = "YAS_ISPREAD_TO_GOVT",
          "wac"        = "AVERAGE_WEIGHTED_COUPON") 
      
      return(df_tmp)
    },
    
    get_data = function(verbose = FALSE){
      
      file_path = file.path(self$raw_path, "RawPrices.parquet")
      if (file.exists(file_path) == TRUE){
        
        if (verbose == TRUE){print("Found ETF Data")}
        df_out <- read_parquet(file = file_path)
        
      }
      else{
        
        if (verbose == TRUE){print("Getting ETF Data")}
        
        df_px         <- self$get_px()
        df_attributes <- self$get_bond_attributes()
        read_path     <- file.path(dirname(self$bbg_px_path), "root/BBGTickers.xlsx")
        
        df_combined <- df_px %>% 
          inner_join(df_attributes, by = c("date", "security"))
        
        df_out  <- read_excel(path = read_path) %>% 
          select(security = Security, desc = Description) %>% 
          inner_join(df_combined, by = "security") %>% 
          group_by(security) %>% 
          arrange(date) %>% 
          mutate(
            security = str_split(string = security, pattern = " ")[[1]][1],
            PX_diff = PX_last - lag(PX_last),
            PX_rtn = PX_diff / lag(PX_last),
            PX_bps = PX_diff / mod_dur) %>% 
          ungroup()
        
        if (verbose == TRUE){print("Saving data\n")}
        write_parquet(df_out, file_path)
        
        }
    
        return(df_out)
    }
    
  ))

df = DataCollect$new()$get_data(verbose = TRUE)

File: root/src/DataCollect.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jan 30 11:07:55 2025

@author: Diego
"""

import os
import numpy as np
import pandas as pd
import yfinance as yf

class DataCollect: 
    
    def __init__(self):
        
        self.dir       = os.path.dirname(os.path.abspath(__file__))  
        self.root_path = os.path.abspath(
            os.path.join(os.path.abspath(
                os.path.join(self.dir, os.pardir)), os.pardir))
        
        self.data_path = os.path.join(self.root_path, "data")
        self.raw_path  = os.path.join(self.data_path, "raw")
        
        if os.path.exists(self.data_path) == False: os.makedirs(self.data_path)
        if os.path.exists(self.raw_path) == False: os.makedirs(self.raw_path)
        
        self.etf_tickers = ["ANGL", "FALN", "JNK", "HYG", "AGG", "LQD", "SJNK"]        
        self.bbg_path    = r"C:\Users\Diego\Desktop\app_prod\BBGData\data"
        self.fund_path   = r"C:\Users\Diego\Desktop\app_prod\BBGData\ETFIndices\BondPricing"
        
        self.bad_variables = ["AVERAGE_WEIGHTED_COUPON"]
        self.clean_window  = 30
        
    def get_raw_px(self, verbose: bool = False) -> pd.DataFrame:
        
        file_path = os.path.join(self.raw_path, "PX.parquet")
        try:
            
            if verbose == True: print("Trying to find raw PX data")
            df_out = pd.read_parquet(path = file_path, engine = "pyarrow")
            if verbose == True: print("Found Data\n")
            
        except: 
        
            if verbose == True: print("Couldn't find data, collecting it now")  
        
            df_out = (yf.download(
                tickers = self.etf_tickers)
                ["Adj Close"].
                reset_index().
                melt(id_vars = ["Date"]).
                dropna().
                rename(columns = {
                    "Date"  : "date",
                    "Ticker": "security",
                    "value" : "PX"}).
                assign(date = lambda x: pd.to_datetime(x.date).dt.date))
            
            if verbose == True: print("Saving data\n")
            df_out.to_parquet(path = file_path, engine = "pyarrow")
            
        return df_out
    
    def get_bond_data(self, verbose: bool = False) -> pd.DataFrame: 
        
        file_path = os.path.join(self.raw_path, "BondFundamentals.parquet")
        try:
            
            if verbose == True: print("Trying to find Bond Fundamentals data")
            df_out = pd.read_parquet(path = file_path, engine = "pyarrow")
            if verbose == True: print("Found Data\n")
            
        except: 
        
            if verbose == True: print("Couldn't find data, collecting it now")  
        
            paths = [os.path.join(
                self.fund_path, ticker + ".parquet")
                for ticker in self.etf_tickers]
            
            renamer = {
                "YAS_YLD_SPREAD"         : "YAS",
                "YAS_BOND_YLD"           : "YLD",
                "YAS_ISPREAD_TO_GOVT"    : "ISPREAD",
                "YAS_MOD_DUR"            : "MOD_DUR"}
            
            df_out = (pd.read_parquet(
                path = paths, engine = "pyarrow").
                assign(security = lambda x: x.security.str.split(" ").str[0]).
                query("variable != @self.bad_variables").
                pivot(index = ["date", "security"], columns = "variable", values = "value").
                rename(columns = renamer).
                reset_index())
            
            if verbose == True: print("Saving data\n")
            df_out.to_parquet(path = file_path, engine = "pyarrow")
            
        return df_out
    
    def _clean(self, df: pd.DataFrame, window: int) -> pd.DataFrame:
        
        df_out = (df.assign(
            diff_val = lambda x: x.value.diff(),
            mean_val    = lambda x: x.diff_val.mean(),
            std_val     = lambda x: x.diff_val.std(),
            z_score     = lambda x: np.abs((x.diff_val - x.mean_val) / x.std_val),
            tmp         = lambda x: np.where(x.z_score > 4, np.nan, x.value),
            replace_val = lambda x: x.tmp.ewm(span = 2, adjust = False).mean()).
            drop(columns = ["diff_val", "mean_val", "std_val", "z_score", "tmp"]).
            rename(columns = {"replace_val": "new_val"}).
            assign(
                mean_val    = lambda x: x.new_val.rolling(window = window).mean(),
                std_val     = lambda x: x.new_val.rolling(window = window).std(),
                z_score     = lambda x: np.abs((x.new_val - x.mean_val) / x.std_val),
                tmp         = lambda x: np.where(x.z_score > 4, np.nan, x.new_val),
                replace_val = lambda x: x.tmp.ewm(span = 2, adjust = False).mean()).
            drop(columns = ["new_val", "mean_val", "std_val", "z_score", "tmp"]))
        
        return df_out
    
    def clean_bond_fundamentals(self, verbose: bool = False) -> pd.DataFrame:
        
        file_path = os.path.join(self.raw_path, "CleanedBondFundamentals.parquet")
        try:
            
            if verbose == True: print("Trying to find Bond Fundamentals data")
            df_out = pd.read_parquet(path = file_path, engine = "pyarrow")
            if verbose == True: print("Found Data\n")
            
        except: 
        
            if verbose == True: print("Couldn't find data, collecting it now")  
        
            df_longer = (self.get_bond_data().melt(
                id_vars = ["date", "security"]))
            
            df_dur = (df_longer.query(
                "variable == 'MOD_DUR'"))
            
            df_clean = (df_longer.query(
                "variable != 'MOD_DUR'").
                groupby(["security", "variable"]).
                apply(self._clean, self.clean_window).
                reset_index(drop = True).
                drop(columns = ["value"]).
                rename(columns = {"replace_val": "value"}))
            
            df_out = (pd.concat([
                df_dur, df_clean]))
        
            if verbose == True: print("Saving data\n")
            df_out.to_parquet(path = file_path, engine = "pyarrow")
            
        return df_out
        
def main() -> None:
           
    DataCollect().get_raw_px(verbose = True)
    DataCollect().get_bond_data(verbose = True)
    DataCollect().clean_bond_fundamentals(verbose = True)
    
if __name__ == "__main__": main()